pkg/kv/kvserver/allocator.go

```diff
@@ -1288,7 +1288,12 @@ func (a *Allocator) TransferLeaseTarget(
 		if preferred[0].StoreID == leaseStoreID {
 			return roachpb.ReplicaDescriptor{}
 		}
-		return preferred[0]
+		// Verify that the preferred replica is eligible to receive the lease.
+		preferred, _ = a.storePool.liveAndDeadReplicas(preferred)
+		if len(preferred) == 1 {
+			return preferred[0]
+		}
+		return roachpb.ReplicaDescriptor{}
```

Comment:
There's a behavior change here that I'm not sure we want in the common case in which there is a single preferred replica that is draining. Prior to this PR, it would fall through and consider moving the lease elsewhere. Now, it will decide not to move the lease. Do we want to preserve the old behavior?

Reply:
Once we are in the `len(preferred) == 1` branch of the code we always return either the preferred lease holder or an empty `ReplicaDescriptor`, this PR does not change that behavior. In your example the draining code calls shedLease, which will pass in `checkTransferLeaseSource=false`, which means the lease holder will get removing from the existing list of candidates before checking for lease preferences, so it falls through that way.